pkg/distributor/distributor.go

```diff
@@ -504,10 +496,28 @@ func New(cfg Config, clientConfig ingester_client.Config, limits *validation.Ove
 		ingestionRateStrategy = newGlobalRateStrategyWithBurstFactor(limits, d)
 	}
 
+	// If this isn't a real distributor that will be accepting writes or if the HA tracker is
+	// disabled, use a no-op implementation. We don't want to require it to be configured when
+	// it's not used due to being disabled or this being a component that doesn't use it (rulers,
+	// queriers).
+	var haTrackerImpl haTracker
+
+	if !canJoinDistributorsRing || !cfg.HATrackerConfig.EnableHATracker {
```

Comment:
Shouldn't the `cfg.HATrackerConfig.EnableHATracker` be part of the HaTracker. Currently as it is if `cfg.HATrackerConfig.EnableHATracker=false`, the newHaTracker will not have any effect 🤔

Reply:
No, this is on purpose. Having the `defaultHaTracker ` need to be functional when it's not enabled causes extra complication. The HA tracker is entirely internal to the distributor so we're using a no-op version when it's not wanted or needed. `newHaTracker()` should only ever be called when a functional HA tracker is needed.

> if we set cfg.HATrackerConfig.EnableHATracker=false, newHaTracker will always create a KV client etc

`newHaTracker()` is never called when `EnableHATracker` is `false`.

Reply:
I see, make sense. Thanks for the explanation